common/src/main/java/io/netty/util/ResourceLeakDetectorFactory.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2016 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.netty.util;
+
+import io.netty.util.internal.ObjectUtil;
+import io.netty.util.internal.PlatformDependent;
+import io.netty.util.internal.SystemPropertyUtil;
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+
+import java.lang.reflect.Constructor;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+
+/**
+ * This static factory should be used to load {@link ResourceLeakDetector}s as needed
+ */
+public abstract class ResourceLeakDetectorFactory {
+    private static final InternalLogger logger = InternalLoggerFactory.getInstance(ResourceLeakDetectorFactory.class);
+
+    private static volatile ResourceLeakDetectorFactory factoryInstance = new ResourceLeakDetectorFactory() {
+        private final String customLeakDetector = AccessController.doPrivileged(new PrivilegedAction<String>() {
+            @Override
+            public String run() {
+                return SystemPropertyUtil.get("io.netty.customResourceLeakDetector");
+            }
+        });
+
+        @Override
+        public <T> ResourceLeakDetector<T> newResourceLeakDetector(final Class<T> resource) {
+            try {
+                if (customLeakDetector != null) {
+                    Class<?> detectorClass = Class.forName(customLeakDetector, true,
+                            PlatformDependent.getSystemClassLoader());
```

Comment:
Whats the problem ? Just catch `Throwable` log it and always use the default impl later on. We do this in several places in netty.